Remove register listener

omerfirmak opened the pull request.

omerfirmak committed:
Validator: Register as soon as node is enrolled

omerfirmak committed:
Manager: Return early from discover task

Using peerLimitReached as a loop condition in discover essentially
results in an infinite loop if used with high enough max_listeners
configuration. This makes updating required_peers impossible
since the timer delegate never returns.

This realies on periodic timer to pick up new addresses discovered
by the AddressDiscoveryTask

omerfirmak committed:
Validator: Clear required_peers on each shuffle

omerfirmak committed:
Remove register[As]Listener

not needed anymore

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -555,7 +555,12 @@ public class Validator : FullNode, API
+        {
+            // Try to register immediately if we are a validator for the next cycle
+            if (this.enroll_man.isEnrolled(block.header.height + 1, &this.utxo_set.peekUTXO))
+                () @trusted { this.network.onRegisterName(); } ();
```
A `@system` function is a function for which we can't prove that there won't be any memory corruption for any combination of parameter. A `@safe` function is one where we can prove it. A `@trusted` lambda is usually used to signify that the call is `@safe` with this set of parameters.

This is easier to demonstrate with C function, let's say `strdup`.
If you do:
```D
void main () @safe
{
    auto ptr = () @trusted { return strdup("Hello World"); }();
}
```

A human can easily look at this, and knowing that string literals are always `\0`-terminated, the call will not result in any memory corruption.

Now when I look at this block of code, I am confused, because `onRegisterName` does not take any parameter.
So the only way that it could be `@system` and proven `@safe` here would be if calling `onRegisterName` was not `@safe` depending on the state of `this`. That would obviously be a very big breach of encapsulation.

So it seems to me that in this case, `onRegisterName` should have been made `@safe`, instead of being wrapped here.

Geod24 commented on source/agora/test/ManyValidators.d:
```diff
@@ -75,6 +80,10 @@ void manyValidators (size_t validators)
             format("Node %s has validator count %s. Expected: %s",
                 idx, node.countActive(Height(GenesisValidatorCycle + 1)), validators)));
 
+    // Wait for nodes to run a discovery task and update their required peers
+    Thread.sleep(3.seconds);
+    network.waitForDiscovery();
```
Shouldn't `waitForDiscovery` handle the sleeping period ?

omerfirmak replied:
Even after block externalization, required_peers in NetworkManager is not updated until the next call to `discover()` which happens periodically in discovery timer. So between the block externalization and the next timer tick, NetworkManager still thinks the network state is complete, even though quorums were shuffled in the last block and makes `waitForDiscovery()` return earlier than it should.

Geod24 replied:
So.... Shouldn't we fix this ?

omerfirmak committed:
Make onRegisterName `@safe`

omerfirmak committed:
FullNode: Refactor getPreimagesRange to getPreimagesFrom

Since some time ago, we only keep the latest preimage in our
storage. So retrieval from a range does not make sense anymore.

This could also make a node to never be able to catch up
if it falls too much behind. Imagine a network at height 2000, and a node
starting from scratch at height 0. Fetching preimages [0, 1024] would
always fail since no preimage in our database will pass the
height < end_height check.

Geod24 commented:
First commit fixes #2455 so thanks for that.
One thing though is that if a node falls too much behind *and* there has been a lot of changes in the validator set, what do we send ? Everything from the looks of it.

Also, nitpick in the commit message:
> if it falls too behind

Much ?

Geod24 commented:
Very nice! Not much to say. The Integration test failed though, I've re-triggered it, let's see if it was spurious, but I doubt so.

Geod24 commented:
Looks like the integration failure is legit @omerfirmak

omerfirmak commented:
> 
> 
> Looks like the integration failure is legit @omerfirmak

There are no errors in the logs, it seems nodes are not shutting down properly.

omerfirmak committed:
Test: address flakey NetworkClient test
